### PR TITLE
Packaging Karoo GP

### PR DESCRIPTION
**Pull request by ilovelinux**
Hi,
this pull request is intended to make Karoo GP installable as Python package, while it keeps working standalone.
When installed as package karoo-gp.py will become an executable command, CSV files will be in the packaged directory and the output will be saved on `~/.karoo_gp/runs/`.

My only question is whether the CSV files need to be accessible to the user or are supposed to be edited. If yes, their actual location isn't great and should be changed to something like `/usr/share/karoo_gp/`.

### CODE REVIEW

**Comment by kstaats** [comment]
Antonio,

Thank you for diving into this. Much appreciated.

I am finally working to upgrade Karoo to the latest Python3 environment 
and libraries. Clearly, Tensorflow has changed to Keras and so I have 
some re-work ahead, or at least I need to import the keras/tf libraries 
in such a way to make them match. Looking into this soon ...

Are you running Karoo as-is? or did you already update Karoo to the 
latest Keras?

Cheers,
kai

**Comment by ilovelinux** [comment]
Kai,

I am running Karoo as is. I actually added a warning in README.md about
compatibility issue (
https://github.com/ilovelinux/karoo_gp/commit/975c20cd60833d1b657bd4a43a10a0a3e0c7a49a)
with Python 3.8+ since Tensorflow < 2 is only available (through PIP) for
Python 3.7 and older. Also, because of this, the package's setup.py
requires Python 3.7 and older.

Cheers,
Antonio

**Comment by ilovelinux** [comment]
The latest version of Scikit Learn (0.24.0, released a few days ago) is no longer compatible with Ubuntu 18.04 because it requires Cython >= 0.28.5 - but the latest version available on Ubuntu 18.04 repositories is 0.26.1.

A possible solution is using Scikit Learn < 0.24.0, or deprecate the support for Ubuntu 18.04.

**Comment by kstaats** [comment]
Antonio,

I have not had time to get Karoo running and fully tested in a new 
Python environment (with Ubuntu 20.4) nor fully test your package. My 
apology. It's been a few weeks of juggling too many things.

Eager to make your effort available to my users, soon.

I forgot, did we resolve this issue of the csv file location?

Cheers,
kai


> this pull request is intended to make Karoo GP installable as Python package, while it keeps working standalone.
> When installed as package karoo-gp.py will become an executable command, CSV files will be in the packaged directory and the output will be saved on `~/.karoo_gp/runs/`.
> 
> My only question is whether the CSV files need to be accessible to the user or are supposed to be edited. If yes, their actual location isn&#39;t great and should be changed to something like `/usr/share/karoo_gp/`.
> 
> You can view, comment on, or merge this pull request online at:
> 
>    https://github.com/kstaats/karoo_gp/pull/21
> 
> -- Commit Summary --
> 
>    * Initial commit
>    * Add warning about compatibility issues.
>    * Prepare directory structure for packaging
>    * Rename to avoid import issues
>    * Minor edits
>    * Initial commit
> 
> -- File Changes --
> 
>      M README.md (2)
>      R karoo-gp.py (6)
>      A karoo_gp/__init__.py (0)
>      R karoo_gp/base_class.py (6)
>      R karoo_gp/files/Iris_dataset/A_little_extra_information_about_the_Iris_data_set.pdf (0)
>      R karoo_gp/files/Iris_dataset/README.txt (0)
>      R karoo_gp/files/Iris_dataset/data_IRIS_setosa-vs-versicolor_3-col.csv (0)
>      R karoo_gp/files/Iris_dataset/data_IRIS_setosa-vs-versicolor_4-col.csv (0)
>      R karoo_gp/files/Iris_dataset/data_IRIS_versicolor-vs-virginica_3-col.csv (0)
>      R karoo_gp/files/Iris_dataset/data_IRIS_versicolor-vs-virginica_4-col.csv (0)
>      R karoo_gp/files/Iris_dataset/data_IRIS_virginica-vs-setosa_3-col.csv (0)
>      R karoo_gp/files/Iris_dataset/data_IRIS_virginica-vs-setosa_4-col.csv (0)
>      R karoo_gp/files/Iris_dataset/iris_data_corrected.csv (0)
>      R karoo_gp/files/Iris_dataset/iris_data_original.csv (0)
>      R karoo_gp/files/coefficients.csv (0)
>      R karoo_gp/files/data_CLASSIFY.csv (0)
>      R karoo_gp/files/data_MATCH.csv (0)
>      R karoo_gp/files/data_PLAY.csv (0)
>      R karoo_gp/files/data_REGRESS.csv (0)
>      R karoo_gp/files/operators_CLASSIFY.csv (0)
>      R karoo_gp/files/operators_MATCH.csv (0)
>      R karoo_gp/files/operators_PLAY.csv (0)
>      R karoo_gp/files/operators_REGRESS.csv (0)
>      R karoo_gp/files/templates/data_iris_2-class.csv (0)
>      R karoo_gp/files/templates/data_iris_3-class.csv (0)
>      R karoo_gp/files/templates/data_kepler.csv (0)
>      R karoo_gp/files/templates/data_power.csv (0)
>      R karoo_gp/files/templates/data_sqrt.csv (0)
>      R karoo_gp/files/templates/data_sum.csv (0)
>      R karoo_gp/files/templates/data_trig.csv (0)
>      R karoo_gp/files/templates/operators_list.txt (0)
>      R karoo_gp/pause.py (0)
>      A requirements.txt (4)
>      A setup.py (39)
> 
> -- Patch Links --
> 
> https://github.com/kstaats/karoo_gp/pull/21.patch
> https://github.com/kstaats/karoo_gp/pull/21.diff
>

File: karoo_gp/base_class.py [hunk 1]
```diff
@@ -367,7 +367,7 @@ def fx_data_load(self, filename):
 		
 		### PART 1 - load the associated data set, operators, operands, fitness type, and coefficients ###
 		# full_path = os.path.realpath(__file__); cwd = os.path.dirname(full_path) # for user Marco Cavaglia
-		cwd = os.getcwd()
+		cwd, _ = os.path.split(__file__)
```
Comment (kstaats) [inline karoo_gp/base_class.py:370]: a) `cwd` is no longer the correct variable name, so we need to change this to `karoo_dir`

b)  use instead `karoo_dir = os.path.dirname(os.path.realpath(__file__))` per Ezio's suggestion

c) update all use-cases of `cwd` with `karoo_dir`

File: setup.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
+import setuptools
+
+with open("README.md", "r") as fh:
+    long_description = fh.read()
+
+setuptools.setup(
+    name="karoo_gp",
+    version="2.3",
```
Comment (kstaats) [inline setup.py:8]: please change the version from `2.3` to `2.4` in all files and Ezio suggests using the __version__ function so as to change it across all files

File: setup.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    package_data={
+        'karoo_gp': ['files/*'],
+    },
+    install_requires=[
+        'numpy',
+        'sklearn',
+        'sympy',
+        'tensorflow<2',
+    ],
+    classifiers=[
```
Comment (kstaats) [inline setup.py:32]: from https://pypi.org/

Topic
 	Scientific/Engineering
		Artificial Intelligence

and then my own additions if we can:
	genetic programming
	gp
	genetic algorithm
	machine learning
	evolutionary computation

Intended Audience
	Developers
	Education
	Science/Research

**Review by kstaats — CHANGES_REQUESTED** [review]

**Pull request merged**